### PR TITLE
Decouple ext scorer init & inference & decoding for the convenience o…

### LINKED ISSUE #117: tools/tune.py感觉没有根据最新的data.py进行更新
在我用tune.py进行调节参数的时候，出现这样的错误：KeyError: u'sequence_offset'，于是我就自己修改了文件，运行。这里提醒一下这个bug,当然，也可能是我自己修改了太多东西，很可能哪里不协调，而官网的本来是协调的。

### LINKED ISSUE #120: Missing num_conv_layers in data_generator (tune.py)
Hi,

I found that there is an issue in tune.py when call DataGenerator function as shown in below.

    data_generator = DataGenerator(
        vocab_filepath=args.vocab_path,
        mean_std_filepath=args.mean_std_path,
        augmentation_config='{}',
        specgram_type=args.specgram_type,
        num_threads=args.num_proc_data,
        keep_transcription_text=True,
        num_conv_layers=args.num_conv_layers)

There is no argument for num_conv_layers in DataGenerator (data_utils/data.py)

Thanks,
Alim

### LINKED ISSUE #121: Decouple ext scorer init & inference & decoding for the convenience of tuning

### PR DESCRIPTION
**Pull request by kuke**
Resolve #121
Fix #120
Fix #117

### CODE REVIEW

File: infer.py [hunk 1]
```diff
@@ -90,18 +90,27 @@ def infer():
...
-        beam_alpha=args.alpha,
-        beam_beta=args.beta,
-        beam_size=args.beam_size,
-        cutoff_prob=args.cutoff_prob,
-        cutoff_top_n=args.cutoff_top_n,
-        vocab_list=vocab_list,
-        language_model_path=args.lang_model_path,
-        num_processes=args.num_proc_bsearch,
-        feeding_dict=data_generator.feeding)
+    if args.decoding_method == "ctc_beam_search":
```
Comment (pkuyym) [inline infer.py]: If `ctc_greedy` doesn't require the external score, I think it's better to move this `if` after line 104

File: model_utils/model.py [hunk 2]
```diff
@@ -173,43 +173,19 @@ def infer_loss_batch(self, infer_data):
         # run inference
         return self._loss_inferer.infer(input=infer_data)
 
-    def infer_batch(self, infer_data, decoding_method, beam_alpha, beam_beta,
-                    beam_size, cutoff_prob, cutoff_top_n, vocab_list,
-                    language_model_path, num_processes, feeding_dict):
-        """Model inference. Infer the transcription for a batch of speech
-        utterances.
+    def infer_probs_batch(self, infer_data, feeding_dict):
```
Comment (pkuyym) [inline model_utils/model.py]: Please keep the naming consistent.  Like below `decode_batch_beam_search`, please change `infer_probs_batch` to `infer_batch_probs`.

File: tools/tune.py [hunk 3]
```diff
@@ -88,76 +83,25 @@ def tune():
...
-        rnn_size=args.rnn_layer_size,
-        use_gru=args.use_gru,
-        share_rnn_weights=args.share_rnn_weights)
+        keep_transcription_text=True)
 
     batch_reader = data_generator.batch_reader_creator(
         manifest_path=args.tune_manifest,
         batch_size=args.batch_size,
         sortagrad=False,
         shuffle_method=None)
```
Comment (pkuyym) [inline tools/tune.py]: Why removed the validation for model path here ?

File: infer.py (hunk 1 above)
Comment (kuke) [reply infer.py]: Done

File: model_utils/model.py (hunk 2 above)
Comment (kuke) [reply model_utils/model.py]: Done

**Pull request merged**